docs/specification.md

```diff
@@ -0,0 +1,127 @@
+on Hyperledger to Tezos. This means investors can have immediate liquidity
+of the corresponding investment. This document explains the framework of the
+smart contract.
+
+To restrict and regulate share transfers `Safelist` contract is used.
+
+## Safelist contract
+
+Only admin can access all non-view functions.
+
+Approximate storage structure:
+```haskell
+data Storage = Storage
+  { owner :: Address
+  , admins :: Set Address
+  , whitelist :: Set Address
+  , blacklist :: Set Address
+  }
+```
+
+Required `Safelist` entrypoint:
+* `addAdmin (address)`
+  * Description: adds address to `admins`.
+  * Constraints: `onlyOwner`.
+* `removeAdmin (address)` onlyOwner.
+  * Description: removes address from `admins`.
+  * Constraints: `onlyOwner`.
+* `isAdmin (view address bool)`
+  * Description: returns `True` if the address is admin.
+* `addToWhitelist (address)` onlyAdmin.
+  * Description: adds address to `whitelist`.
+  * Constraints: `onlyAdmin`.
+* `removeFromWhitelist (address)` onlyAdmin.
+  * Description: removes address from `whitelist`.
+  * Constraints: `onlyAdmin`.
+* `isWhitelisted (view address bool)`
+  * Description: returns `True` if the address is whitelisted.
+* `addToBlacklist (address)` onlyAdmin.
+  * Description: adds address to `blacklist`.
+  * Constraints: `onlyAdmin`.
+* `removeFromBlacklist (address)` onlyAdmin.
+  * Description: removes address from `blacklist`.
+  * Constraints: `onlyAdmin`.
+* `isBlacklisted (view address bool)`
+  * Description: returns `True` if the address is blacklisted.
+
+Additionaly, it would be nice to have the following entrypoint:
+* `checkConstraints ([address, constraint])`.
+  * Description: here `constraint` is property that needs to be checked for given
+  address. Can be one of the following:
+  `isWhitelisted`, `isNonBlacklisted`.
+  This entrypoint doesn't do anything in case all constraints are satisfied
+  and fails with an error otherwise.
+
+This contract implements regulatory service for the **Holdings** conract.
+
+## Holdings contract
+
+Approximate storage structure:
+```haskell
+data Storage = Storage
+  { name :: MText
+  , symbol :: MText
+  , id :: MText
+  , safelistAddress :: Maybe (TAddress Safelist.Storage)
+  , owner :: Address
+  , admin :: Address
+  , ledger :: BigMap Address (Map Address Natural, Natural)
+  , paused :: Bool
+  , transferable :: Bool
+  , totalMinted :: Natural
+  , totalBurned :: Natural
+  }
+```
+
+Required `Holdings` entrypoints:
+* `setName (string :newName)`
+  * Description: updates `name` field.
+  * Constraints: `onlyAdmin` `isNotPaused`.
+* `setSymbol (string :newSymbol)`
+  * Description: updates `symbol` field.
+  * Constraints: `onlyAdmin` `isNotPaused`.
+* `setSafelistAddress ((option address) :mbNewSafelistAddress)`
+  * Description: updates `safelistAddress` field, which is used for safelist contract calls.
+  * Constraints: `onlyOwner`.
+* `setAdminAddress (address)`
+  * Description: updates `admin` field.
```

Comment:
Probably one day auditors will come and say that it's better to use a two-step process. Maybe let's stick to it from the beginning?